I2C Keyboard support for the Cardputer ADV

SpaceMonkeyAlfa opened the pull request:
# Summary
This PR implements the I2C keyboard logic for the userinput library. It also includes a modified settings app.

# Why the new approach?
I've had a little trouble getting the other two PRs from @carnitashredder and @TheOddCell to work on my Cardputer ADV. This PR contains my own _keys.py implementation, based on the M5Stack Cardputer library for Arduino IDE. 
As it is based on the official M5Stack Cardputer Library, this implementation works a little differently to the others, so I thought I'd share it in case it's useful.

@blounolito mentions in @TheOddCell's PR that some apps / scripts crash for them. In my own implementation, I had this issue with the built in settings app, but managed to fix it by modifying the settings.py script.

 It seems that I2C processing is a little more memory intensive then the old keyboard system for the Cardputer, and the settings app's more complex menus / draw calls fill up memory. After some tweaks (loading the display libraries first, more memory cleanup), I got the settings app working. I've tested a couple other apps from GetApps, and they seem to work fine,  _but any other more complex apps will likely have similar memory issues to settings._

I’ve tested this with my modified Settings app and several GetApps applications, and it is currently stable. I hope someone finds this PR helpful!

SpaceMonkeyAlfa committed:
Add definition.yml for CARDPUTER_ADV

SpaceMonkeyAlfa committed:
New _keys.py script for input handling

Implement key mapping and I2C handling for Cardputer ADV.

SpaceMonkeyAlfa committed:
Memory changes for settings app

SpaceMonkeyAlfa committed:
Update _keys.py

SpaceMonkeyAlfa committed:
Update top comment

echo-lalia commented:
Thanks so much for your help with this! 😊

In regards to your comments about memory issues: did you test it out with just the plain .py files, or did you try compiling any of it to either .mpy or into a firmware .bin? Because either of those options will free up a lot of memory (The original Cardputer has memory issues with the current version of MicroHydra when it's not compiled, too).

echo-lalia commented:
Just a quick update on the functionality of the builds I posted above: The user "r0da" from the MH Discord installed the .bin I shared on their Cardputer Adv, and says apps seem to work totally fine, except for a lack of audio.

So, I think we can safely discard the modified settings app, but we'll still need to make changes to the audio to fully support the Adv :)

SpaceMonkeyAlfa committed:
Delete patched settings.py

Building a micropython .bin is sufficient to fix the memory issues with opening "demanding" apps. This patch for settings.py is no longer needed
